### PR TITLE
Upgrade v4

### PR DESCRIPTION
**Pull request by pasotee**
### Describe the purpose of your pull request

- Upgrade Unreal Engine SDK to support the [cpp-sdk](https://github.com/configcat/cpp-sdk) [v4 release[(https://github.com/configcat/cpp-sdk/releases/tag/v4.0.0)

### CODE REVIEW

File: Source/ConfigCat/Private/ConfigCatSubsystem.cpp [hunk 1]
```diff
@@ -217,20 +211,20 @@ void UConfigCatSubsystem::ForceRefresh()
 	ConfigCatClient->forceRefresh();
 }
 
-void UConfigCatSubsystem::SetDefaultUser(const FConfigCatUser& User)
+void UConfigCatSubsystem::SetDefaultUser(const UConfigCatUserWrapper* User)
 {
 	if (!ensure(ConfigCatClient))
```
Comment (kp-cat) [inline Source/ConfigCat/Private/ConfigCatSubsystem.cpp:216]: Could you please create an EnsureConfigCatClient() function where this log message can be printed:
```
bool EnsureConfigCatCline()
{
	if (ensure(ConfigCatClient))
	{
                return true;
        }
        
	UE_LOG(LogConfigCat, Warning, TEXT("Trying to access the ConfigCatClient before initialization or after shutdown."));
	return false;
}
```

In every function we can use this EnsureConfigCatClient() to check the client easier:

```
if (!EnsureConfigCatClient())
{
        return;
}
```

File: Source/ConfigCatWrappers/Private/ConfigCatEvaluationWrapper.cpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
+﻿// Fill out your copyright notice in the Description page of Project Settings.
```
Comment (kp-cat) [inline Source/ConfigCatWrappers/Private/ConfigCatEvaluationWrapper.cpp:1]: What is the standard? Is this line needed here?

File: Source/ConfigCatWrappers/Private/ConfigCatPercentageOptionWrapper.cpp [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+uint8 UConfigCatPercentageOptionWrapper::GetPercentage() const
+{
+	return PercentageOption.percentage;
+}
+
+FString UConfigCatPercentageOptionWrapper::GetVariationId() const
+{
+	if(PercentageOption.variationId)
```
Comment (kp-cat) [inline Source/ConfigCatWrappers/Private/ConfigCatPercentageOptionWrapper.cpp:21]:
```suggestion
	if (PercentageOption.variationId)
```

File: Source/ConfigCatWrappers/Private/ConfigCatEvaluationWrapper.cpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	{
+		return UConfigCatTargetingRuleWrapper::CreateTargetingRule(EvaluationDetails->matchedTargetingRule.value());
+	}
+
+	return {};
+}
+
+UConfigCatPercentageOptionWrapper* UConfigCatEvaluationWrapper::GetMatchedPercentageOption() const
+{
+	if(EvaluationDetails && EvaluationDetails->matchedPercentageOption)
```
Comment (kp-cat) [inline Source/ConfigCatWrappers/Private/ConfigCatEvaluationWrapper.cpp:111]:
```suggestion
	if (EvaluationDetails && EvaluationDetails->matchedPercentageOption)
```

File: Source/ConfigCatWrappers/Private/ConfigCatEvaluationWrapper.cpp [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	if(EvaluationDetails && EvaluationDetails->errorException)
+	{
+		return UTF8_TO_TCHAR(unwrap_exception_message(EvaluationDetails->errorException).c_str());
+	}
+	return {};
+}
+
+UConfigCatTargetingRuleWrapper* UConfigCatEvaluationWrapper::GetMatchedTargetingRule() const
+{
+	if(EvaluationDetails && EvaluationDetails->matchedTargetingRule)
```
Comment (kp-cat) [inline Source/ConfigCatWrappers/Private/ConfigCatEvaluationWrapper.cpp:101]:
```suggestion
	if (EvaluationDetails && EvaluationDetails->matchedTargetingRule)
```

File: Source/ConfigCatWrappers/Private/ConfigCatEvaluationWrapper.cpp [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+	{
+		return UTF8_TO_TCHAR(EvaluationDetails->errorMessage.value().c_str());
+	}
+
+	return {};
+}
+
+FString UConfigCatEvaluationWrapper::GetException() const
+{
+	if(EvaluationDetails && EvaluationDetails->errorException)
```
Comment (kp-cat) [inline Source/ConfigCatWrappers/Private/ConfigCatEvaluationWrapper.cpp:92]:
``` suggestion
	if (EvaluationDetails && EvaluationDetails->errorException)
```

File: Source/ConfigCatWrappers/Private/ConfigCatSettingsWrapper.cpp [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+EConfigCatSettingTypeWrapper UConfigCatSettingWrapper::GetType() const
+{
+	return static_cast<EConfigCatSettingTypeWrapper>(Setting.type);
+}
+
+FString UConfigCatSettingWrapper::GetPercentageOptionsAttribute() const
+{
+	if(Setting.percentageOptionsAttribute)
```
Comment (kp-cat) [inline Source/ConfigCatWrappers/Private/ConfigCatSettingsWrapper.cpp:30]:
```suggestion
	if (Setting.percentageOptionsAttribute)
```

File: Source/ConfigCatWrappers/Private/ConfigCatSettingsWrapper.cpp [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+	{
+		return UTF8_TO_TCHAR(Setting.percentageOptionsAttribute.value().c_str());
+	}
+
+	return {};
+}
+
+FString UConfigCatSettingWrapper::GetVariationId() const
+{
+	if(Setting.variationId)
```
Comment (kp-cat) [inline Source/ConfigCatWrappers/Private/ConfigCatSettingsWrapper.cpp:40]:
```suggestion
	if (Setting.variationId)
```

File: Source/ConfigCatWrappers/Private/ConfigCatSettingsWrapper.cpp [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+		Result.Add(UConfigCatTargetingRuleWrapper::CreateTargetingRule(Rule));
+	}
+	
+	return Result;
+}
+
+TArray<UConfigCatPercentageOptionWrapper*> UConfigCatSettingWrapper::GetPercentageOptions() const
+{
+	TArray<UConfigCatPercentageOptionWrapper*> Result;
+	for(const auto& Option : Setting.percentageOptions)
```
Comment (kp-cat) [inline Source/ConfigCatWrappers/Private/ConfigCatSettingsWrapper.cpp:67]:
```suggestion
	for (const auto& Option : Setting.percentageOptions)
```

File: Source/ConfigCatWrappers/Private/ConfigCatTargetingRuleWrapper.cpp [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+	{
+		return EConfigCatUserComparator::Invalid;
+	}
+	
+	return static_cast<EConfigCatUserComparator>(UserCondition.comparator);
+}
+
+FString UConfigCatUserConditionWrapper::GetStringComparisonValue() const
+{
+	// UserCondition.comparisonValue;
```
Comment (kp-cat) [inline Source/ConfigCatWrappers/Private/ConfigCatTargetingRuleWrapper.cpp:33]: I think you can implement it like this: 
```suggestion
	const auto comparisonValuePtr = get_if<string>(&UserCondition.comparisonValue);
	if (comparisonValuePtr)
	{
		return UTF8_TO_TCHAR(comparisonValuePtr->c_str());
	}
```

File: Source/ConfigCatWrappers/Private/ConfigCatTargetingRuleWrapper.cpp [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+	return {};
+}
+
+double UConfigCatUserConditionWrapper::GetNumberComparisonValue() const
+{
+	// UserCondition.comparisonValue;
```
Comment (kp-cat) [inline Source/ConfigCatWrappers/Private/ConfigCatTargetingRuleWrapper.cpp:39]:
```suggestion
	const auto comparisonValuePtr = get_if<double>(&UserCondition.comparisonValue);
	if (comparisonValuePtr)
	{
		return *comparisonValuePtr;
	}
```

File: Source/ConfigCatWrappers/Private/ConfigCatTargetingRuleWrapper.cpp [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+	return {};
+}
+
+TArray<FString> UConfigCatUserConditionWrapper::GetStringArrayComparisonValue() const
+{
+	// UserCondition.comparisonValue;
+	return {};
```
Comment (kp-cat) [inline Source/ConfigCatWrappers/Private/ConfigCatTargetingRuleWrapper.cpp:46]:
```suggestion
	TArray<FString> Result;
	const auto comparisonValuePtr = get_if<vector<string>>(&UserCondition.comparisonValue);
	if (comparisonValuePtr)
	{
		for (const auto& value : *comparisonValuePtr)
		{
			Result.Emplace(UTF8_TO_TCHAR(value.c_str()));
		}
	}
	
	return Result;
```

File: Source/ConfigCatWrappers/Private/ConfigCatUserWrapper.cpp [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+		const auto TimeSinceEpoch = std::get<date_time_t>(*Attribute).time_since_epoch().count();
+		return FDateTime::FromUnixTimestamp(TimeSinceEpoch);
+	}
+
+	return {};
+}
+
+TArray<FString> UConfigCatUserWrapper::GetStringArrayAttribute(const FString& Key) const
+{
+	if (HasStringAttribute(Key))
```
Comment (kp-cat) [inline Source/ConfigCatWrappers/Private/ConfigCatUserWrapper.cpp:80]:
```suggestion
	if (HasStringArrayAttribute(Key))
```

File: Source/ConfigCatWrappers/Private/ConfigCatEvaluationWrapper.cpp (hunk 2 above)
Comment (pasotee) [reply Source/ConfigCatWrappers/Private/ConfigCatEvaluationWrapper.cpp]: that is an oversight on my side. I have adjusted it with the proper copyright.

File: Source/ConfigCat/Private/ConfigCatSubsystem.cpp (hunk 1 above)
Comment (pasotee) [reply Source/ConfigCat/Private/ConfigCatSubsystem.cpp]: That is a great idea. I added it now, thanks.

File: Source/ConfigCatWrappers/Private/ConfigCatTargetingRuleWrapper.cpp (hunk 10 above)
Comment (pasotee) [reply Source/ConfigCatWrappers/Private/ConfigCatTargetingRuleWrapper.cpp]: Thanks for spotting those and for the suggestions, I didn't know they were a thing.

However, I have to implement the `Has...Value` functions as well, so I will be using that to check the values.

File: Source/ConfigCatWrappers/Private/ConfigCatTargetingRuleWrapper.cpp (hunk 11 above)
Comment (pasotee) [reply Source/ConfigCatWrappers/Private/ConfigCatTargetingRuleWrapper.cpp]: Function is now implemented

File: Source/ConfigCatWrappers/Private/ConfigCatTargetingRuleWrapper.cpp (hunk 12 above)
Comment (pasotee) [reply Source/ConfigCatWrappers/Private/ConfigCatTargetingRuleWrapper.cpp]: Function is now implemented

**Comment by pasotee** [comment]
> Thanks @pasotee for the PR. Could you please fix the configcat dependency in the CI yml configs?
> 
> ```
> name: Plugin CD
> 
> on:
>   workflow_dispatch:
>   push:
>     tags:
>       - 'v*'
> 
> env:
>   VCPKG_COMMIT_HASH: "59852bb5bd0779788e5fc95e50d2048ae6c9399d"
> 
> jobs:
>   build-win64:
>     runs-on: windows-latest
> 
>     steps:
>       - uses: actions/checkout@v2
> 
>       - name: Install via vcpkg
>         run: |
>           git clone https://github.com/microsoft/vcpkg build/vcpkg
>           git -C build/vcpkg checkout ${{ env.VCPKG_COMMIT_HASH }}
>           ./build/vcpkg/bootstrap-vcpkg.bat
>           ./build/vcpkg/vcpkg install configcat[core] --triplet x64-windows
> 
>       - uses: actions/upload-artifact@v3
>         with:
>           name: Win64
>           path: |
>             ./build/vcpkg/installed/*/lib/configcat.lib
>             ./build/vcpkg/installed/*/lib/hash-library.lib
> ```

This is awesome. I think it would benefit both the CI & CD pipelines, so I've added it to both

**Review by kp-cat — APPROVED** [review]

**Pull request merged**